# 🤗 Use Notus on inference endpoints to create a legal preference dataset

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/argilla-io/distilabel/blob/main/docs/tutorials/pipeline-notus-instructions-preferences-legal.ipynb) [![Open Source in Github](https://img.shields.io/badge/github-view%20source-black.svg)](https://github.com/argilla-io/distilabel/blob/main/docs/tutorials/pipeline-notus-instructions-preferences-legal.ipynb)


In this tutorial, you will learn how to use the Notus model on Inference Endpoints to create a legal preference dataset based on RAG instructions from the European AI Act. A full end-to-end example of how to use distilabel to leverage LLMs!

[distilabel](https://github.com/argilla-io/distilabel) is an AI Feedback (AIF) framework that can generate and label datasets using LLMs, and can be used for many different use cases. Implemented with robustness, efficiency and scalability in mind, it allows anyone to build their synthetic datasets that can be used in many different scenarios. This tutorial shows an end-to-end example in which we will create a model expert in the new AI Act, to which we can make different types of questions and requests.

The LLM model that we will fine-tune for this is [Notus 7B](https://argilla.io/blog/notus7b/), a fine-tuned version of Zephyr 7B that uses Direct Preference Optimization (DPO) and AIF techniques to outperform its foundation model in several benchmarks, and is completely open-source.

This tutorial includes the following steps:

- Defining a custom generating task for a `distilabel` pipeline.
- Creating a RAG pipeline using Haystack for the EU AI Act.
- Generating an instruction dataset with `SelfInstructTask`.
- Generating a preference dataset using an `UltraFeedback` text quality task.

You can use the Open in Colab button at the top of this page. This option allows you to run the notebook directly on Google Colab. Don't forget to change the runtime type to GPU for faster model training and inference.

## Introduction

Let's start by installing the required depencies to run distilabel, Argilla and the rest of the packages used in the tutorial; most notably, Haystack.


In [1]:
%pip install argilla distilabel "farm-haystack[preprocessing]" pip install "distilabel[hf-inference-endpoints]"


[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


### Running Argilla

For this tutorial, you can use Argilla to visualize and annotate the different datasets created by distilabel. There are two main options for deploying and running Argilla:

**Deploy Argilla on Hugging Face Spaces:** If you want to run tutorials with external notebooks (e.g., Google Colab) and you have an account on Hugging Face, you can deploy Argilla on Spaces with a few clicks:

[![deploy on spaces](https://huggingface.co/datasets/huggingface/badges/raw/main/deploy-to-spaces-lg.svg)](https://huggingface.co/new-space?template=argilla/argilla-template-space)

For details about configuring your deployment, check the [official Hugging Face Hub guide](https://huggingface.co/docs/hub/spaces-sdks-docker-argilla).

**Launch Argilla using Argilla's quickstart Docker image**: This is the recommended option if you want [Argilla running on your local machine](../../getting_started/quickstart.ipynb). Note that this option will only let you run the tutorial locally and not with an external notebook service.

For more information on deployment options, please check the Deployment section of the documentation.

<div class="alert alert-info">

Tip

This tutorial is a Jupyter Notebook. There are two options to run it:

- Use the Open in Colab button at the top of this page. This option allows you to run the notebook directly on Google Colab. Don't forget to change the runtime type to GPU for faster model training and inference.
- Download the .ipynb file by clicking on the View source link at the top of the page. This option allows you to download the notebook and run it on your local machine or on a Jupyter notebook tool of your choice.
</div>


### Import dependencies

The main dependencies for this tutorial are distilabel for creating the synthetic datasets and Argilla for visualizing and annotating these datasets, and also for fine-tuning our model. The package [Haystack](https://haystack.deepset.ai/) is used to creates batches from the original PDF document we want to create our datasets from.


In [2]:
import os
from typing import Dict

import argilla as rg

from distilabel.llm import InferenceEndpointsLLM
from distilabel.pipeline import Pipeline, pipeline
from distilabel.tasks import TextGenerationTask, SelfInstructTask, Prompt

from datasets import Dataset
from haystack.nodes import PDFToTextConverter, PreProcessor

/Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


If you are running Argilla using the Docker quickstart image or Hugging Face Spaces, you need to init the Argilla client with the URL and API_KEY:


In [3]:
# Replace api_url with the url to your HF Spaces URL if using Spaces
# Replace api_key if you configured a custom API key
rg.init(
    api_url="https://ignacioct-argilla.hf.space",
    api_key="owner.apikey",
    workspace="admin",
)

Additionally, we need to provide our HuggingFace and OpenAI accest token. To later instatiate an `InferenceEndpointsLLM` object, we need to pass as parameters the HF Inference Endpoint name and the HF namespace. One very convenient way to do so is also through environment variables.


In [4]:
os.environ["HF_TOKEN"] = "hf_"
os.environ["HF_INFERENCE_ENDPOINT_NAME"] = "aws-notus-7b-v1-3184"
os.environ["HF_NAMESPACE"] = "argilla"
os.environ["OPENAI_API_KEY"] = "sk-"

## Setting up an inference endpoint with Notus

Inference endpoints are a solution, managed by Hugging Face, to easily deploy any Transformer-like model. They are built from models on the Hugging Face Hub. Inference endpoints are really handy for making inference on LLMs without the hastle of trying to run the models locally. In this tutorial, we will use inference endpoints to generate text using our Notus model, as part of the `distilabel` workflow. The endpoint of choice has a [Notus 7B instance](https://ui.endpoints.huggingface.co/argilla/endpoints/aws-notus-7b-v1-4052) running.

### Defining a custom generating task for a distilabel pipeline

To kickstart this tutorial, let's see how to set up and endpoint for our Notus model. It's not part of the end-to-end example we'll see later, but an example of how to connect to a Hugging Face endpoint and a test of the `distilabel` pipeline.

Let's dive into this quick example of how to use an inference endpoint. We have prepared an easy `TextGenerationTask` to ask question to the model, in a very similar way as we talk with the LLMs using chatbots. First, we define a class for the question-answering task, with functions showing `distilabel` how the model should generate the prompts, parse the input and the output, etc.


In [5]:
class QuestionAnsweringTask(TextGenerationTask):
    def generate_prompt(self, question: str) -> str:
        return Prompt(
            system_prompt=self.system_prompt,
            formatted_prompt=question,
        ).format_as(
            "llama2"
        )  # type: ignore

    def parse_output(self, output: str) -> Dict[str, str]:
        return {"answer": output.strip()}

    @property
    def input_args_names(self) -> list[str]:
        return ["question"]

    @property
    def output_args_names(self) -> list[str]:
        return ["answer"]

`llm` is an object of the `InferenceEndpointsLLM` class, and by using it we can start generating answers to question using the `llm.generate()` method.


In [6]:
llm = InferenceEndpointsLLM(
    endpoint_name_or_model_id=os.getenv("HF_INFERENCE_ENDPOINT_NAME"),  # type: ignore
    endpoint_namespace=os.getenv("HF_NAMESPACE"),  # type: ignore
    token=os.getenv("HF_TOKEN") or None,
    task=QuestionAnsweringTask(),
)

[01/17/24 15:57:05] INFO     INFO:distilabel:Using Dedicated Inference Endpoint          ]8;id=904534;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/llm/huggingface/inference_endpoints.py\inference_endpoints.py]8;;\:]8;id=241931;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/llm/huggingface/inference_endpoints.py#158\158]8;;\

With the `InferenceEndpointsLLM` object defined with the endpoint information and the Task, we can go ahead and start generating text. Let's ask this LLM what's, for example, the second most populated city in Denmark. The answer should be Aarhus.


In [7]:
generation = llm.generate(
    [{"question": "What's the second most populated city in Denmark?"}]
)
generation[0][0]["parsed_output"]["answer"]

'The second most populated city in Denmark is Aarhus, with a population of around 340,000 people. It is located on the east coast of Jutland, and is known for its vibrant cultural scene, beautiful beaches, and historic landmarks. Aarhus is also home to Aarhus University, one of the largest universities in Scandinavia.'

The endpoint is working correctly! We have succesfully set up a custom generating task for a `distilabel` pipeline.


## Creating a RAG pipeline using Haystack for the European AI Act

For this end-to-end example, we would like to create an expert model capable of answering question and filling up information about the new AI Act promoted by the European Union, which is the first regulation on artificial intelligence. As part of its digital strategy, the EU wants to regulate artificial AI to ensure better conditions for the development and use of this innovative technology. This act is a regulatory framework for AI, with different risk levels meaning more or less regulation. They are the world's first rules on AI.

This RAG pipeline that we want to create downloads the PDF file, converts it to plain text and preprocess it, creating batches that we can feed `distilabel` to start creating instructions from it. Let's see this first part of the pipeline and get the input data. Note that this RAG part of the pipeline is not based on an active pipeline based queries or semantic properties, but a more brute-force approach in which we download the PDF and preprocess its contents.

### Downloading the AI Act PDF

Firstly, we need to download the PDF document itself. We'll place it in our working directory, if it's not there already.


In [8]:
%%bash

if [ ! -f "The-AI-Act.pdf" ]; then
    wget -q https://artificialintelligenceact.eu/wp-content/uploads/2021/08/The-AI-Act.pdf
fi

Once we have it in our working directory, we can use Haystack's Converter and Pipeline features to extract the textual data, clean it and divide it in different batches. Afterwards, these batches will be used to start creating synthetic instructions.


In [9]:
# The converter turns the PDF into text we can process easily
converter = PDFToTextConverter(remove_numeric_tables=True, valid_languages=["en"])

# Preprocessing pipelines can have several steps.
# Ours clean empty lines, header, footers and whitespaces
# and split the text into 150-char long batches, respecting
# where the sentences naturally end and begin.
preprocessor = PreProcessor(
    clean_empty_lines=True,
    clean_whitespace=True,
    clean_header_footer=True,
    split_by="word",
    split_length=150,
    split_respect_sentence_boundary=True,
)

doc = converter.convert(file_path="The-AI-Act.pdf", meta=None)[0]
docs = preprocessor.process([doc])
print(f"Documents: 1\nBatches: {len(docs)}")

pdftotext version 4.04 [www.xpdfreader.com]
Copyright 1996-2022 Glyph & Cog, LLC
Preprocessing:   0%|          | 0/1 [00:00<?, ?docs/s]

[01/17/24 15:57:11] WARNING  WARNING:haystack.nodes.preprocessor.preprocessor:We found one or   ]8;id=709027;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/haystack/nodes/preprocessor/preprocessor.py\preprocessor.py]8;;\:]8;id=953552;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/haystack/nodes/preprocessor/preprocessor.py#516\516]8;;\
                             more sentences whose split count is higher than the split length.                     

Preprocessing: 100%|██████████| 1/1 [00:00<00:00,  5.02docs/s]

Documents: 1
Batches: 355


Let's take a quick look at the batches we just generated.


In [10]:
inputs = [doc.content for doc in docs]
inputs[0][0:500]

'EN EN\nEUROPEAN\nCOMMISSION\nProposal for a\nREGULATION OF THE EUROPEAN PARLIAMENT AND OF THE COUNCIL\nLAYING DOWN HARMONISED RULES ON ARTIFICIAL INTELLIGENCE\n(ARTIFICIAL INTELLIGENCE ACT) AND AMENDING CERTAIN UNION\nLEGISLATIVE ACTS\x0cEN\nEXPLANATORY MEMORANDUM\n1. CONTEXT OF THE PROPOSAL\n1.1. Reasons for and objectives of the proposal\nThis explanatory memorandum accompanies the proposal for a Regulation laying down\nharmonised rules on artificial intelligence (Artificial Intelligence Act). Artificial Int'

The document has been correctly batched, from one big document to 355 strings, 150-character long at maximum. This list of strings can now be used as input to generate a instruction dataset using `distilabel`.


## Generating instructions with SelfInstructTask

With out Inference Endpoint up and running, we should be able to generate instructions with distilabel. These instructions, made by the LLM through our endpoint, will form an instruction dataset, with instructions created from the data we just extracted.

For this example, we are using a subset of 50 batches generated in the section above, to be gentle on performance.


In [11]:
instructions_dataset = Dataset.from_dict({"input": inputs[0:50]})

instructions_dataset

Dataset({
    features: ['input'],
    num_rows: 50
})

With the `SelfInstructTask` class we can generate a Self-Instruct specitification for building the prompts, as done in the [Self-Instruct paper](https://arxiv.org/abs/2212.10560). `distilabel` will start from human-made input, in this case, the batches we created from the AI Act pdf, and it will generate instructions based on it. These instructions can then be reviewed using Argilla to keep the best ones.

An application description can be passed as a parameter to specify the behaviour of the model; we want a model capable of answering our questions about the AI Act.


In [12]:
instructions_task = SelfInstructTask(
    application_description="A assistant that can answer questions about the AI Act made by the European Union."
)

Let's now define a generator, passing the `SelfInstructTask` object, and create a `Pipeline` object.


In [13]:
instructions_generator = InferenceEndpointsLLM(
    endpoint_name_or_model_id=os.getenv("HF_INFERENCE_ENDPOINT_NAME"),  # type: ignore
    endpoint_namespace=os.getenv("HF_NAMESPACE"),  # type: ignore
    token=os.getenv("HF_TOKEN") or None,
    task=instructions_task,
)

instructions_pipeline = Pipeline(generator=instructions_generator)

                    INFO     INFO:distilabel:Using Dedicated Inference Endpoint          ]8;id=745703;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/llm/huggingface/inference_endpoints.py\inference_endpoints.py]8;;\:]8;id=15723;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/llm/huggingface/inference_endpoints.py#158\158]8;;\

Our pipeline is ready to be used to generate instructions. Let's do it!


In [14]:
generated_instructions = instructions_pipeline.generate(
    dataset=instructions_dataset, num_generations=1, batch_size=8
)

                    INFO     INFO:distilabel:Executing dry-run...                                   ]8;id=342706;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=836205;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#747\747]8;;\

                    INFO     INFO:distilabel:Processing batch 1 of 1...                             ]8;id=722178;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=568063;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 1...                       ]8;id=393405;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=754079;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

/Users/ignacio/Documents/recognai/distilabel/src/distilabel/llm/huggingface/inference_endpoints.py:220: UserWarning: No `prompt_format` has been specified and no `default_format` is set, so the prompt will be concatenated with a line-break and no specific formatting by default.
  prompts = self._generate_prompts(inputs, default_format=None)


Flattening the indices: 100%|██████████| 1/1 [00:00<00:00, 87.40 examples/s]


[01/17/24 15:57:15] INFO     INFO:distilabel:Dry-run executed with no issues. Starting the actual   ]8;id=794794;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=932210;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#749\749]8;;\
                             generation...                                                                         

/Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/rich/live.py:231: UserWarning: 
install "ipywidgets" for Jupyter support
  warnings.warn('install "ipywidgets" for Jupyter support')

                    INFO     INFO:distilabel:Processing batch 1 of 7...                             ]8;id=376361;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=460325;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 1...                       ]8;id=887489;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=703943;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 15:57:50] INFO     INFO:distilabel:Processing batch 2 of 7...                             ]8;id=535488;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=106534;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 2...                       ]8;id=546797;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=524639;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 15:58:28] INFO     INFO:distilabel:Processing batch 3 of 7...                             ]8;id=13636;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=630603;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 3...                       ]8;id=5424;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=205078;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 15:59:04] INFO     INFO:distilabel:Processing batch 4 of 7...                             ]8;id=902213;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=403054;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 4...                       ]8;id=399531;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=231476;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 15:59:42] INFO     INFO:distilabel:Processing batch 5 of 7...                             ]8;id=81406;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=446016;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 5...                       ]8;id=253861;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=846292;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:00:18] INFO     INFO:distilabel:Processing batch 6 of 7...                             ]8;id=245420;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=370918;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 6...                       ]8;id=717813;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=736126;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:00:55] INFO     INFO:distilabel:Processing batch 7 of 7...                             ]8;id=320728;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=315354;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 7...                       ]8;id=911120;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=598389;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

Saving the dataset (1/1 shards): 100%|██████████| 50/50 [00:00<00:00, 15537.91 examples/s]


[01/17/24 16:01:04] INFO     INFO:distilabel:Final dataset saved at                                 ]8;id=668240;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=275414;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#658\658]8;;\
                             /Users/ignacio/Documents/recognai/distilabel/docs/tutorials/ckpt                      

The pipeline has succesfully generated instructions given the topics and the behaviour passed as input. Let's gather all those instructions and see how the look.


In [15]:
instructions = []
for generations in generated_instructions["instructions"]:
    for generation in generations:
        instructions.extend(generation)

print(f"Number of generated instructions: {len(instructions)}")

for instruction in instructions[:5]:
    print(instruction)

Number of generated instructions: 178
What are the reasons for and objectives of the proposal for a Regulation laying down harmonised rules on artificial intelligence?
How can artificial intelligence improve prediction, optimise operations and resource allocation, and personalise service delivery?
What benefits can artificial intelligence bring to the European economy and society as a whole?
How can the use of artificial intelligence support socially and environmentally beneficial outcomes?
What are the high-impact sectors that require AI action according to the AI Act by the European Union?


These initial intructions form our instruction dataset. Following the human-in-the-loop approach, we should push the instructions to Argilla to visualize them and be able to rank them in terms of quality. Those annotations are essential to make quality data, ensuring a better performance of the final model. Nevertheless, this step is optional.


### Pushing the instruction dataset to Argilla to visualize and annotate.

Let's take a quick look at the instructions generated by `SelfInstructTask`.


In [16]:
generated_instructions[0]

{'input': 'EN EN\nEUROPEAN\nCOMMISSION\nProposal for a\nREGULATION OF THE EUROPEAN PARLIAMENT AND OF THE COUNCIL\nLAYING DOWN HARMONISED RULES ON ARTIFICIAL INTELLIGENCE\n(ARTIFICIAL INTELLIGENCE ACT) AND AMENDING CERTAIN UNION\nLEGISLATIVE ACTS\x0cEN\nEXPLANATORY MEMORANDUM\n1. CONTEXT OF THE PROPOSAL\n1.1. Reasons for and objectives of the proposal\nThis explanatory memorandum accompanies the proposal for a Regulation laying down\nharmonised rules on artificial intelligence (Artificial Intelligence Act). Artificial Intelligence\n(AI) is a fast evolving family of technologies that can bring a wide array of economic and\nsocietal benefits across the entire spectrum of industries and social activities. By improving\nprediction, optimising operations and resource allocation, and personalising service delivery,\nthe use of artificial intelligence can support socially and environmentally beneficial outcomes\nand provide key competitive advantages to companies and the European economy. ',
 

For each input, i.e., each batch of the AI Act pdf file, we have a generator prompt, with general guidelines on how to behave, as well as the application description parameter. 4 instructions per input have been generated.

Now it's the perfect time to upload the instruction dataset to Argilla, review it and manually annotate it.

In [17]:
instructions_rg_dataset = generated_instructions.to_argilla()
instructions_rg_dataset[0]

FeedbackRecord(fields={'input': 'EN EN\nEUROPEAN\nCOMMISSION\nProposal for a\nREGULATION OF THE EUROPEAN PARLIAMENT AND OF THE COUNCIL\nLAYING DOWN HARMONISED RULES ON ARTIFICIAL INTELLIGENCE\n(ARTIFICIAL INTELLIGENCE ACT) AND AMENDING CERTAIN UNION\nLEGISLATIVE ACTS\x0cEN\nEXPLANATORY MEMORANDUM\n1. CONTEXT OF THE PROPOSAL\n1.1. Reasons for and objectives of the proposal\nThis explanatory memorandum accompanies the proposal for a Regulation laying down\nharmonised rules on artificial intelligence (Artificial Intelligence Act). Artificial Intelligence\n(AI) is a fast evolving family of technologies that can bring a wide array of economic and\nsocietal benefits across the entire spectrum of industries and social activities. By improving\nprediction, optimising operations and resource allocation, and personalising service delivery,\nthe use of artificial intelligence can support socially and environmentally beneficial outcomes\nand provide key competitive advantages to companies and the 

In [18]:
instructions_rg_dataset.push_to_argilla(name=f"notus_AI_instructions")

[01/17/24 16:01:34] INFO     INFO:argilla.client.feedback.dataset.local.mixins:✓ Dataset succesfully  ]8;id=448902;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py\mixins.py]8;;\:]8;id=337536;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py#281\281]8;;\
                             pushed to Argilla                                                                     

                    INFO     INFO:argilla.client.feedback.dataset.local.mixins:RemoteFeedbackDataset( ]8;id=311207;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py\mixins.py]8;;\:]8;id=119682;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py#282\282]8;;\
                                id=aaa43d72-467a-4e4c-b37c-40d50115fd89                                            
                                name=notus_AI_instructions                                                         
                                workspace=Workspace(id=5f137d29-7fee-4eca-bdaa-554b93dbd928,                       
                             name=admin, inserted_at=2024-01-17 14:47:19.686508,                                   
                             updated_at=2024-01-17 14:47:19.686508)                                                
                                url=https://ignacioct-argilla.hf.space/dataset/aaa43d72-467a-4e4c-b37              
                             c-40d50115fd89/annotation-mode                                                        
                                fields=[RemoteTextField(id=UUID('4b449269-feb1-4da4-baa1-04140b909c93              
                             '), client=None, name='input', title='input', required=True,                          
                             type='text', use_markdown=True),                                                      
                             RemoteTextField(id=UUID('679301f2-e78e-4668-a66d-83eb34cae684'),                      
                             client=None, name='instruction', title='instruction', required=True,                  
                             type='text', use_markdown=False)]                                                     
                                questions=[RemoteRatingQuestion(id=UUID('9836dcf6-f962-4826-bf00-c192              
                             6e44e4e1'), client=None, name='instruction-rating', title='How would you              
                             rate the generated instruction?', description=None, required=True,                    
                             type='rating', values=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10])]                               
                                guidelines=None                                                                    
                                metadata_properties=[RemoteIntegerMetadataProperty(id=UUID('0a4ba15c-              
                             19a0-4216-a748-613a661881fb'), client=<httpx.Client object at                         
                             0x2a0b5ffd0>, name='length-input', title='length-input',                              
                             visible_for_annotators=True, type='integer', min=None, max=None),                     
                             RemoteIntegerMetadataProperty(id=UUID('ab3c2cdd-162c-4a03-bfb4-e29be98a5              
                             9d3'), client=<httpx.Client object at 0x2a0b5ffd0>,                                   
                             name='length-instruction', title='length-instruction',                                
                             visible_for_annotators=True, type='integer', min=None, max=None)]                     
                             )                                                                                     

RemoteFeedbackDataset(
   id=aaa43d72-467a-4e4c-b37c-40d50115fd89
   name=notus_AI_instructions
   workspace=Workspace(id=5f137d29-7fee-4eca-bdaa-554b93dbd928, name=admin, inserted_at=2024-01-17 14:47:19.686508, updated_at=2024-01-17 14:47:19.686508)
   url=https://ignacioct-argilla.hf.space/dataset/aaa43d72-467a-4e4c-b37c-40d50115fd89/annotation-mode
   fields=[RemoteTextField(id=UUID('4b449269-feb1-4da4-baa1-04140b909c93'), client=None, name='input', title='input', required=True, type='text', use_markdown=True), RemoteTextField(id=UUID('679301f2-e78e-4668-a66d-83eb34cae684'), client=None, name='instruction', title='instruction', required=True, type='text', use_markdown=False)]
   questions=[RemoteRatingQuestion(id=UUID('9836dcf6-f962-4826-bf00-c1926e44e4e1'), client=None, name='instruction-rating', title='How would you rate the generated instruction?', description=None, required=True, type='rating', values=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10])]
   guidelines=None
   metadata_properties=[R

In the Argilla UI, each tuple input-instruction is visualized individually, and can be individually annotated.

![](../assets/tutorials-assets/instrucion_dataset_ui.png)


## Generate a Preference Dataset using an Ultrafeedback text quality task.

Once we have our instruction dataset, we are going to create a preference dataset through the UltraFeedback text quality task. This is a type of task used in NLP used to evaluate the quality of text generated; our goal is to provide detailed feedback on the quality of the generated text, beyond a binary label.

Our `pipeline()` method allows us to create a `Pipeline` instance with the provided LLMs for a given task, which is useful whenever you want to use a pre-defined or custom `Pipeline` for a given task. We will specify our task and subtask, the generator we want to use (in this case, one based in a Text Generator Task) and our OpenAI API key.


In [19]:
preference_pipeline = pipeline(
    "preference",
    "instruction-following",
    generator=InferenceEndpointsLLM(
        endpoint_name_or_model_id=os.getenv("HF_INFERENCE_ENDPOINT_NAME"),  # type: ignore
        endpoint_namespace=os.getenv("HF_NAMESPACE", None),
        task=TextGenerationTask(),
        max_new_tokens=256,
        num_threads=2,
        temperature=0.3,
    ),
    max_new_tokens=256,
    num_threads=2,
    openai_api_key=os.getenv("OPENAI_API_KEY"),
    temperature=0.0,
)

[01/17/24 16:01:56] INFO     INFO:distilabel:Using Dedicated Inference Endpoint          ]8;id=711646;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/llm/huggingface/inference_endpoints.py\inference_endpoints.py]8;;\:]8;id=681291;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/llm/huggingface/inference_endpoints.py#158\158]8;;\

[01/17/24 16:01:57] INFO     INFO:distilabel:Since no `labeller` was provided, `OpenAILLM` will be  ]8;id=577857;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=419363;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#838\838]8;;\
                             used as the default labeller with `UltraFeedback`.                                    

We also need to retrieve our instruction dataset from Argilla, as it will be the input of this pipeline.


In [20]:
remote_dataset = rg.FeedbackDataset.from_argilla(
    "notus_AI_instructions", workspace="admin"
)
instructions_dataset = remote_dataset.pull(max_records=100)  # get first 100 records

instructions_dataset = instructions_dataset.format_as("datasets")
instructions_dataset

Dataset({
    features: ['input', 'instruction', 'instruction-rating', 'instruction-rating-suggestion', 'instruction-rating-suggestion-metadata', 'external_id', 'metadata'],
    num_rows: 100
})

In [21]:
instructions_dataset[0]

{'input': 'EN EN\nEUROPEAN\nCOMMISSION\nProposal for a\nREGULATION OF THE EUROPEAN PARLIAMENT AND OF THE COUNCIL\nLAYING DOWN HARMONISED RULES ON ARTIFICIAL INTELLIGENCE\n(ARTIFICIAL INTELLIGENCE ACT) AND AMENDING CERTAIN UNION\nLEGISLATIVE ACTS\x0cEN\nEXPLANATORY MEMORANDUM\n1. CONTEXT OF THE PROPOSAL\n1.1. Reasons for and objectives of the proposal\nThis explanatory memorandum accompanies the proposal for a Regulation laying down\nharmonised rules on artificial intelligence (Artificial Intelligence Act). Artificial Intelligence\n(AI) is a fast evolving family of technologies that can bring a wide array of economic and\nsocietal benefits across the entire spectrum of industries and social activities. By improving\nprediction, optimising operations and resource allocation, and personalising service delivery,\nthe use of artificial intelligence can support socially and environmentally beneficial outcomes\nand provide key competitive advantages to companies and the European economy.',
 '

Before generating the text based on our instructions, we need to mingle a little bit with the dataset. From the previous section, we still have our old input, the batches from the PDF. We have to change that to the instructions that we generated.

In [22]:
instructions_dataset = instructions_dataset.rename_columns({"input": "context", "instruction": "input"})

Now, let's build a dataset by using the pipeline we just created, and the topics from which our instructions were generated.


In [26]:
preference_dataset = preference_pipeline.generate(
    instructions_dataset,  # type: ignore
    num_generations=2,
    batch_size=8,
    display_progress_bar=True,
)

[01/17/24 16:08:48] INFO     INFO:distilabel:Executing dry-run...                                   ]8;id=531459;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=391513;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#747\747]8;;\

                    INFO     INFO:distilabel:Processing batch 1 of 1...                             ]8;id=76706;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=25527;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 1...                       ]8;id=764404;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=306352;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:08:57] INFO     INFO:distilabel:Calling labeller for batch 1...                        ]8;id=302797;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=706574;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

Flattening the indices: 100%|██████████| 1/1 [00:00<00:00, 142.45 examples/s]


[01/17/24 16:09:00] INFO     INFO:distilabel:Dry-run executed with no issues. Starting the actual   ]8;id=426576;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=546207;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#749\749]8;;\
                             generation...                                                                         

                    INFO     INFO:distilabel:Processing batch 1 of 13...                            ]8;id=592957;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=127850;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 1...                       ]8;id=16568;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=312740;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:10:15] INFO     INFO:distilabel:Calling labeller for batch 1...                        ]8;id=306129;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=344462;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 2 of 13...                            ]8;id=128624;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=420613;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 2...                       ]8;id=262709;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=404455;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:11:13] INFO     INFO:distilabel:Calling labeller for batch 2...                        ]8;id=380521;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=541333;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 3 of 13...                            ]8;id=766442;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=205843;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 3...                       ]8;id=333383;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=671471;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:11:58] INFO     INFO:distilabel:Calling labeller for batch 3...                        ]8;id=368480;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=661522;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 4 of 13...                            ]8;id=618383;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=624387;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 4...                       ]8;id=858039;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=302881;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:14:03] INFO     INFO:distilabel:Calling labeller for batch 4...                        ]8;id=605711;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=955426;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 5 of 13...                            ]8;id=823069;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=863334;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 5...                       ]8;id=758721;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=752801;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:15:16] INFO     INFO:distilabel:Calling labeller for batch 5...                        ]8;id=304015;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=833779;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 6 of 13...                            ]8;id=237115;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=259764;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 6...                       ]8;id=770854;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=791679;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:16:29] INFO     INFO:distilabel:Calling labeller for batch 6...                        ]8;id=435751;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=879736;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 7 of 13...                            ]8;id=685910;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=30237;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 7...                       ]8;id=693428;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=32700;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:17:42] INFO     INFO:distilabel:Calling labeller for batch 7...                        ]8;id=188126;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=163561;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 8 of 13...                            ]8;id=607349;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=54426;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 8...                       ]8;id=574477;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=238638;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:18:56] INFO     INFO:distilabel:Calling labeller for batch 8...                        ]8;id=223308;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=833076;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 9 of 13...                            ]8;id=82850;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=152505;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 9...                       ]8;id=409759;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=438182;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:20:10] INFO     INFO:distilabel:Calling labeller for batch 9...                        ]8;id=876536;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=745910;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 10 of 13...                           ]8;id=30109;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=409060;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 10...                      ]8;id=137809;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=71171;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:21:16] INFO     INFO:distilabel:Calling labeller for batch 10...                       ]8;id=426300;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=544288;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 11 of 13...                           ]8;id=499666;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=837678;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 11...                      ]8;id=607144;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=471299;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:22:29] INFO     INFO:distilabel:Calling labeller for batch 11...                       ]8;id=736449;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=945366;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 12 of 13...                           ]8;id=306047;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=341173;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 12...                      ]8;id=594380;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=703369;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:23:43] INFO     INFO:distilabel:Calling labeller for batch 12...                       ]8;id=811335;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=422050;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

                    INFO     INFO:distilabel:Processing batch 13 of 13...                           ]8;id=905726;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=620580;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#569\569]8;;\

                    INFO     INFO:distilabel:Calling generator for batch 13...                      ]8;id=870735;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=769986;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#573\573]8;;\

[01/17/24 16:24:10] INFO     INFO:distilabel:Calling labeller for batch 13...                       ]8;id=311983;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=300042;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#604\604]8;;\

Saving the dataset (1/1 shards): 100%|██████████| 100/100 [00:00<00:00, 13422.63 examples/s]


[01/17/24 16:24:17] INFO     INFO:distilabel:Final dataset saved at                                 ]8;id=650848;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py\pipeline.py]8;;\:]8;id=23407;file:///Users/ignacio/Documents/recognai/distilabel/src/distilabel/pipeline.py#658\658]8;;\
                             /Users/ignacio/Documents/recognai/distilabel/docs/tutorials/ckpt                      

Let's take a look at an instance of the preference dataset


In [27]:
preference_dataset[0]

{'context': 'EN EN\nEUROPEAN\nCOMMISSION\nProposal for a\nREGULATION OF THE EUROPEAN PARLIAMENT AND OF THE COUNCIL\nLAYING DOWN HARMONISED RULES ON ARTIFICIAL INTELLIGENCE\n(ARTIFICIAL INTELLIGENCE ACT) AND AMENDING CERTAIN UNION\nLEGISLATIVE ACTS\x0cEN\nEXPLANATORY MEMORANDUM\n1. CONTEXT OF THE PROPOSAL\n1.1. Reasons for and objectives of the proposal\nThis explanatory memorandum accompanies the proposal for a Regulation laying down\nharmonised rules on artificial intelligence (Artificial Intelligence Act). Artificial Intelligence\n(AI) is a fast evolving family of technologies that can bring a wide array of economic and\nsocietal benefits across the entire spectrum of industries and social activities. By improving\nprediction, optimising operations and resource allocation, and personalising service delivery,\nthe use of artificial intelligence can support socially and environmentally beneficial outcomes\nand provide key competitive advantages to companies and the European economy.',


### Upload the preference dataset to Argilla to annotate.

Once our preference dataset has been correctly generated, the Argilla UI is the best tool at our disposal to visualize it and annotate it. As for the instruction dataset, we just have to convert it to an Argilla Feedback Dataset, and push it to Argilla.


In [28]:
# Uploading the Preference Dataset
preference_rg_dataset = preference_dataset.to_argilla()

# Adding the context as a metadata property in the new Feedback dataset, as this
# information will be useful later.
for record_feedback, record_huggingface in zip(
    preference_rg_dataset, preference_dataset
):
    record_feedback.metadata["context"] = record_huggingface["context"]

preference_rg_dataset.push_to_argilla(name=f"notus_AI_preference")

[01/17/24 16:24:34] INFO     INFO:argilla.client.feedback.dataset.local.mixins:✓ Dataset succesfully  ]8;id=697317;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py\mixins.py]8;;\:]8;id=627828;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py#281\281]8;;\
                             pushed to Argilla                                                                     

                    INFO     INFO:argilla.client.feedback.dataset.local.mixins:RemoteFeedbackDataset( ]8;id=439572;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py\mixins.py]8;;\:]8;id=557405;file:///Users/ignacio/Documents/recognai/distilabel/.venv/lib/python3.10/site-packages/argilla/client/feedback/dataset/local/mixins.py#282\282]8;;\
                                id=a1a3fc1e-d263-4890-8a16-c598e5e52311                                            
                                name=notus_AI_preference                                                           
                                workspace=Workspace(id=5f137d29-7fee-4eca-bdaa-554b93dbd928,                       
                             name=admin, inserted_at=2024-01-17 14:47:19.686508,                                   
                             updated_at=2024-01-17 14:47:19.686508)                                                
                                url=https://ignacioct-argilla.hf.space/dataset/a1a3fc1e-d263-4890-8a1              
                             6-c598e5e52311/annotation-mode                                                        
                                fields=[RemoteTextField(id=UUID('86e1dd49-08a6-4560-8d12-c477b1c7fe7f              
                             '), client=None, name='input', title='input', required=True,                          
                             type='text', use_markdown=True),                                                      
                             RemoteTextField(id=UUID('6ef957e3-c870-4385-8eed-be36135245b8'),                      
                             client=None, name='generations-1', title='generations-1', required=True,              
                             type='text', use_markdown=True),                                                      
                             RemoteTextField(id=UUID('55bf8bb1-2506-4499-9adb-1de1753116bc'),                      
                             client=None, name='generations-2', title='generations-2', required=True,              
                             type='text', use_markdown=True)]                                                      
                                questions=[RemoteRatingQuestion(id=UUID('ccc6bd83-8d5e-426b-a496-67e0              
                             2d52abdd'), client=None, name='generations-1-rating', title="What's the               
                             rating for generations-1?", description=None, required=True,                          
                             type='rating', values=[1, 2, 3, 4, 5]),                                               
                             RemoteRatingQuestion(id=UUID('c99ace66-6652-42ce-8913-74b4046b216f'),                 
                             client=None, name='generations-2-rating', title="What's the rating for                
                             generations-2?", description=None, required=True, type='rating',                      
                             values=[1, 2, 3, 4, 5]),                                                              
                             RemoteTextQuestion(id=UUID('f0b7d095-d4d6-44ca-8e15-f08a706b87b3'),                   
                             client=None, name='rating-rationale', title="What's the rationale behind              
                             each rating?", description=None, required=True, type='text',                          
                             use_markdown=False)]                                                                  
                                guidelines=None                                                                    
                                metadata_properties=[RemoteIntegerMetadataProperty(id=UUID('c5e8e738-              
                             6127-449e-83b0-b2d34e11a14b'), client=<httpx.Client objec

RemoteFeedbackDataset(
   id=a1a3fc1e-d263-4890-8a16-c598e5e52311
   name=notus_AI_preference
   workspace=Workspace(id=5f137d29-7fee-4eca-bdaa-554b93dbd928, name=admin, inserted_at=2024-01-17 14:47:19.686508, updated_at=2024-01-17 14:47:19.686508)
   url=https://ignacioct-argilla.hf.space/dataset/a1a3fc1e-d263-4890-8a16-c598e5e52311/annotation-mode
   fields=[RemoteTextField(id=UUID('86e1dd49-08a6-4560-8d12-c477b1c7fe7f'), client=None, name='input', title='input', required=True, type='text', use_markdown=True), RemoteTextField(id=UUID('6ef957e3-c870-4385-8eed-be36135245b8'), client=None, name='generations-1', title='generations-1', required=True, type='text', use_markdown=True), RemoteTextField(id=UUID('55bf8bb1-2506-4499-9adb-1de1753116bc'), client=None, name='generations-2', title='generations-2', required=True, type='text', use_markdown=True)]
   questions=[RemoteRatingQuestion(id=UUID('ccc6bd83-8d5e-426b-a496-67e02d52abdd'), client=None, name='generations-1-rating', title="What's 

In the Argilla UI, we can see the input (an instruction), and the two generations that the LLM created out of it.

![](../assets/tutorials-assets/preference_dataset_ui.png)


## Conclusions

To conclude, we have gone through an end-to-end example of distilabel. We've set up an Inference Endpoint, defined a distilabel pipeline that extracts information from a PDF, created and manually reviewed the instruction and preference dataset created from that input. The final preference dataset is perfect for fine-tuning, and you can easily do this using the ArgillaTrainer from Argilla. Have a look at these resources if you want to go further:

- [Train a Model with ArgillaTrainer](https://docs.argilla.io/en/latest/tutorials_and_integrations/tutorials/feedback/end2end_examples/train-model-006.html)
- [Ⓜ️ Finetuning LLMs as chat assistants: Supervised Finetuning on Mistral 7B](https://docs.argilla.io/en/latest/tutorials_and_integrations/tutorials/feedback/training-llm-mistral-sft.html)
- [🌠 Improving RAG by Optimizing Retrieval and Reranking Models](https://docs.argilla.io/en/latest/tutorials_and_integrations/tutorials/feedback/fine-tuning-sentencesimilarity-rag.html)
